File: app/src/main/java/com/kaltura/kflow/presentation/main/MainFragment.java
package com.kaltura.kflow.presentation.main;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kaltura.kflow.R;
import com.kaltura.kflow.presentation.AnonymousLoginFragment;
import com.kaltura.kflow.presentation.EpgFragment;
import com.kaltura.kflow.presentation.FavoritesFragment;
import com.kaltura.kflow.presentation.LiveTvFragment;
import com.kaltura.kflow.presentation.LoginFragment;
import com.kaltura.kflow.presentation.MediaPageFragment;
import com.kaltura.kflow.presentation.productPrice.ProductPriceFragment;
import com.kaltura.kflow.presentation.RegistrationFragment;
import com.kaltura.kflow.presentation.GetVodFragment;
import com.kaltura.kflow.presentation.SearchFragment;
import com.kaltura.kflow.presentation.SettingsFragment;
import com.kaltura.kflow.presentation.subscription.SubscriptionFragment;
import com.kaltura.kflow.presentation.transactionHistory.TransactionHistoryFragment;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import static com.kaltura.kflow.presentation.main.Feature.ANONYMOUS_LOGIN;
import static com.kaltura.kflow.presentation.main.Feature.EPG;
import static com.kaltura.kflow.presentation.main.Feature.FAVORITES;
import static com.kaltura.kflow.presentation.main.Feature.LIVE;
import static com.kaltura.kflow.presentation.main.Feature.LOGIN;
import static com.kaltura.kflow.presentation.main.Feature.MEDIA_PAGE;
import static com.kaltura.kflow.presentation.main.Feature.PRODUCT_PRICE;
import static com.kaltura.kflow.presentation.main.Feature.REGISTRATION;
import static com.kaltura.kflow.presentation.main.Feature.SEARCH;
import static com.kaltura.kflow.presentation.main.Feature.SETTINGS;
import static com.kaltura.kflow.presentation.main.Feature.SUBSCRIPTION;
import static com.kaltura.kflow.presentation.main.Feature.TRANSACTION_HISTORY;
import static com.kaltura.kflow.presentation.main.Feature.VOD;

/**
 * Created by alex_lytvynenko on 11/18/18.
 */
public class MainFragment extends Fragment implements FeatureAdapter.OnFeatureClickListener {

    private Feature[] mFeatures = {LOGIN, ANONYMOUS_LOGIN, REGISTRATION, VOD, EPG, LIVE, FAVORITES,
            SEARCH, MEDIA_PAGE, SUBSCRIPTION, PRODUCT_PRICE, TRANSACTION_HISTORY, SETTINGS};

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_main, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initList();
    }

    private void initList() {
        RecyclerView list = getView().findViewById(R.id.list);
        list.setHasFixedSize(true);
        list.setLayoutManager(new LinearLayoutManager(requireContext()));
        list.addItemDecoration(new DividerItemDecoration(requireContext(), LinearLayoutManager.VERTICAL));
        FeatureAdapter adapter = new FeatureAdapter(mFeatures, this);
        list.setAdapter(adapter);
    }

    @Override
    public void onFeatureClicked(Feature feature) {
        switch (feature) {
            case LOGIN:
                pushFragment(new LoginFragment());
                break;
            case ANONYMOUS_LOGIN:
                pushFragment(new AnonymousLoginFragment());
                break;
            case REGISTRATION:
                pushFragment(new RegistrationFragment());
                break;
            case VOD:
                pushFragment(new GetVodFragment());
                break;
            case EPG:
                pushFragment(new EpgFragment());
                break;
            case LIVE:
                pushFragment(new LiveTvFragment());
                break;
            case FAVORITES:
                pushFragment(new FavoritesFragment());
                break;
            case SEARCH:
                pushFragment(new SearchFragment());
                break;
            case MEDIA_PAGE:
                pushFragment(new MediaPageFragment());
                break;
            case SUBSCRIPTION:
                pushFragment(new SubscriptionFragment());
                break;
            case PRODUCT_PRICE:
                pushFragment(new ProductPriceFragment());
                break;
            case TRANSACTION_HISTORY:
                pushFragment(new TransactionHistoryFragment());
                break;
            case SETTINGS:
                pushFragment(new SettingsFragment());
                break;
        }
    }

    private void pushFragment(Fragment fragment) {
        requireActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.container, fragment)
                .addToBackStack(null)
                .commit();
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/main/MainFragment.kt
package com.kaltura.kflow.presentation.main

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.doOnPreDraw
import androidx.dynamicanimation.animation.SpringAnimation
import androidx.dynamicanimation.animation.SpringForce
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.*
import com.kaltura.kflow.R
import com.kaltura.kflow.databinding.FragmentMainBinding
import com.kaltura.kflow.presentation.extension.isTv
import com.kaltura.kflow.presentation.extension.navigateWithExtras

/**
 * Created by alex_lytvynenko on 11/18/18.
 */
class MainFragment : Fragment(R.layout.fragment_main) {

    private val features = arrayOf(Feature.LOGIN, Feature.LOGIN_APP_TOKEN, Feature.WORK_WITH_KS,
            Feature.ANONYMOUS_LOGIN, Feature.REGISTRATION, Feature.COLLECTIONS, Feature.VOD,
            Feature.CONTINUE_WATCHING, Feature.EPG, Feature.LIVE, Feature.FAVORITES, Feature.SEARCH,
            Feature.KEEP_ALIVE, Feature.MEDIA_PAGE, Feature.SUBSCRIPTION, Feature.PRODUCT_PRICE,
            Feature.CHECK_RECEIPT, Feature.TRANSACTION_HISTORY, Feature.RECORDINGS, Feature.BOOKMARK,
            Feature.IOT, Feature.DEVICE_MANAGEMENT,Feature.SNS,Feature.REMINDERS, Feature.SETTINGS)

    private lateinit var rotationAnimation: SpringAnimation
    private var isDragging = false
    private var _binding: FragmentMainBinding? = null
    private val binding get() = _binding!!
    private val scrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            if (isDragging) {
                rotationAnimation.cancel()
                binding.kaltura.rotation += -(dy.toFloat() / 2)
            }
        }

        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            isDragging = newState == RecyclerView.SCROLL_STATE_DRAGGING
            if (newState == RecyclerView.SCROLL_STATE_SETTLING) rotationAnimation.start()
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = FragmentMainBinding.inflate(inflater,container,false)
        val view = binding.root
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
//        _binding = FragmentMainBinding.bind(view)
        postponeEnterTransition()
        view.doOnPreDraw { startPostponedEnterTransition() }
        initSpringAnimation()
        initList()
    }

    private fun initList() {
        binding.list.setHasFixedSize(true)
        binding.list.layoutManager = GridLayoutManager(requireContext(), if (isTv()) 4 else 2)
        val adapter = FeatureAdapter(features)
        adapter.clickListener = { feature, image, title ->
            navigateWithExtras(when (feature) {
                Feature.LOGIN -> MainFragmentDirections.navigateToLogin()
                Feature.LOGIN_APP_TOKEN -> MainFragmentDirections.navigateToAppToken()
                Feature.WORK_WITH_KS -> MainFragmentDirections.navigateToKs()
                Feature.ANONYMOUS_LOGIN -> MainFragmentDirections.navigateToAnonymousLogin()
                Feature.REGISTRATION -> MainFragmentDirections.navigateToRegistration()
                Feature.COLLECTIONS -> MainFragmentDirections.navigateToCollections()
                Feature.VOD -> MainFragmentDirections.navigateToVod()
                Feature.CONTINUE_WATCHING -> MainFragmentDirections.navigateToContinueWatching()
                Feature.EPG -> MainFragmentDirections.navigateToEpg()
                Feature.LIVE -> MainFragmentDirections.navigateToLiveTv()
                Feature.FAVORITES -> MainFragmentDirections.navigateToFavorites()
                Feature.SEARCH -> MainFragmentDirections.navigateToSearch()
                Feature.KEEP_ALIVE -> MainFragmentDirections.navigateToKeepAlive()
                Feature.MEDIA_PAGE -> MainFragmentDirections.navigateToMediaPage()
                Feature.SUBSCRIPTION -> MainFragmentDirections.navigateToSubscription()
                Feature.PRODUCT_PRICE -> MainFragmentDirections.navigateToProductPrice()
                Feature.CHECK_RECEIPT -> MainFragmentDirections.navigateToCheckReceipt()
                Feature.TRANSACTION_HISTORY -> MainFragmentDirections.navigateToTransactionHistory()
                Feature.RECORDINGS -> MainFragmentDirections.navigateToRecordings()
                Feature.BOOKMARK -> MainFragmentDirections.navigateToBookmark()
                Feature.IOT -> MainFragmentDirections.navigateToIot()
                Feature.DEVICE_MANAGEMENT -> MainFragmentDirections.navigateToDeviceManagement()
                Feature.REMINDERS -> MainFragmentDirections.navigateToReminders()
                Feature.SNS -> MainFragmentDirections.navigateToSns()
                Feature.SETTINGS -> MainFragmentDirections.navigateToSettings()
            }, image, title)
        }
        binding.list.adapter = adapter
        binding.list.addOnScrollListener(scrollListener)
    }

    private fun initSpringAnimation() {
        rotationAnimation = SpringAnimation(binding.kaltura, SpringAnimation.ROTATION).apply {
            spring = SpringForce(0f).apply {
                stiffness = SpringForce.STIFFNESS_LOW
                dampingRatio = SpringForce.DAMPING_RATIO_MEDIUM_BOUNCY
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding.list.removeOnScrollListener(scrollListener)
        _binding = null
    }
}